src/neural/backends/network_onnx.cc

```diff
@@ -133,46 +216,178 @@ class OnnxNetwork : public Network {
+        output_tensors_data_[i] =
+            malloc(max_batch_size * output_tensors_step_[i] * data_size);
+      }
+      input_tensor_data_device_ = input_tensor_data_;
+      for (int i = 0; i < outputs_size; i++) {
+        output_tensors_data_device_[i] = output_tensors_data_[i];
+      }
+      memory_info_ =
+          Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
   }
-  if (network_->mlh_head_ != -1) {
-    output_tensors_step_[network_->mlh_head_] = 1;
-    output_tensors_data_[network_->mlh_head_] =
-        std::vector<DataType>(network_->max_batch_size_);
+}
+
+OnnxNetwork::~OnnxNetwork() {
+#ifdef USE_ONNX_CUDART
+  if (provider_ == OnnxProvider::TRT || provider_ == OnnxProvider::CUDA) {
+    ReportCUDAErrors(cudaStreamDestroy(compute_stream_));
+    ReportCUDAErrors(cudaStreamDestroy(upload_stream_));
+    ReportCUDAErrors(cudaStreamDestroy(download_stream_));
   }
+#endif
+}
+
+template <typename DataType>
+OnnxComputation<DataType>::OnnxComputation(OnnxNetwork* network)
+    : network_(network) {
+  inputs_outputs_ = network_->GetInputsOutputs();
+}
+
+template <typename DataType>
+OnnxComputation<DataType>::~OnnxComputation() {
+  network_->ReleaseInputsOutputs(std::move(inputs_outputs_));
+}
+
+void AsDataType(float x, float* y) { *y = x; }
+void AsDataType(float x, Ort::Float16_t* y) {
+  uint16_t tmp = FP32toFP16(x);
+  std::memcpy(reinterpret_cast<uint16_t*>(y), &tmp, sizeof(uint16_t));
+}
+void AsDataType(float x, Ort::BFloat16_t* y) {
+  uint16_t tmp = FP32toBF16(x);
+  std::memcpy(reinterpret_cast<uint16_t*>(y), &tmp, sizeof(uint16_t));
 }
 
 template <typename DataType>
 void OnnxComputation<DataType>::AddInput(InputPlanes&& input) {
+#ifdef USE_ONNX_CUDART
+  if (network_->provider_ == OnnxProvider::CUDA ||
+      network_->provider_ == OnnxProvider::TRT) {
+    assert(input.size() == kInputPlanes);
+    uint64_t* masks =
+        static_cast<uint64_t*>(inputs_outputs_->input_tensor_data_) +
+        input_size_ * kInputPlanes;
+    uint64_t* mask_end =
+        static_cast<uint64_t*>(inputs_outputs_->input_tensor_data_) +
+        network_->max_batch_size_ * kInputPlanes;
+    DataType* values =
+        reinterpret_cast<DataType*>(mask_end) + input_size_ * kInputPlanes;
+    for (size_t i = 0; i < kInputPlanes; i++) {
+      masks[i] = input[i].mask;
+      DataType value;
+      AsDataType(input[i].value, &value);
+      values[i] = value;
+    }
+    input_size_++;
+    if (input_size_ > network_->max_batch_size_) {
+      throw Exception("NN input exceeds max batch size of " +
+                      std::to_string(network_->max_batch_size_) + ".");
+    }
+    return;
+  }
+#endif
   raw_input_.emplace_back(input);
   if (raw_input_.size() > network_->max_batch_size_) {
     throw Exception("NN input exceeds max batch size of " +
                     std::to_string(network_->max_batch_size_) + ".");
   }
 }
+template <typename DataType>
+int OnnxComputation<DataType>::GetBatchSize() const {
```

Comment:
`GetBatchSize()` is a remnant of the old API that isn't currently used (except in the python bindings that need updating anyway). I think it would be a lot cleaner to use `input_size_` everywhere, and it will make it easy to get rid of `GetBatchSize()` in the near future.